bindings/bind_py_structure_manager.cc

```diff
@@ -458,6 +458,132 @@ namespace rascal {
         py::call_guard<py::gil_scoped_release>());
   }
 
+  /**
+   * Bind a getter for the feature matrix comming from BlockSparseProperty.
+   *
+   * it will return a dictionary with keys->atomic number vals->feature mat
+   * the feature matrices size is (n_centers, inner_size)
+   */
+  template <class Calculator, class ManagerCollection_t,
+            class ManagerCollectionBinder>
+  void bind_sparse_feature_matrix_getter(
+      ManagerCollectionBinder & manager_collection) {
+    manager_collection.def(
+        "get_sparse_feature_matrix",
+        [](ManagerCollection_t & managers, Calculator & calculator) {
+          using Manager_t = typename ManagerCollection_t::Manager_t;
+          using Prop_t = typename Calculator::template Property_t<Manager_t>;
+          using Keys_t = typename Prop_t::Keys_t;
+          using VecMap_t = Eigen::Map<math::Vector_t>;
+
+          auto property_name{managers.get_calculator_name(calculator, false)};
+
+          auto && property_ =
+              managers[0]->template get_property_ref<Prop_t>(property_name);
+          // assume inner_size is consistent for all managers
+          int inner_size{property_.get_nb_comp()};
+          auto n_rows{managers.get_number_of_elements(calculator, false)};
+
+          // holder for the feature matrices to put in feature_dict
+          math::Matrix_t features{};
+
+          features.resize(n_rows, inner_size);
+          features.setZero();
+
+          // get  the keys for the dictionary
+          Keys_t all_keys{};
+          for (auto & manager : managers) {
+            auto && property =
+                manager->template get_property_ref<Prop_t>(property_name);
+            auto keys = property.get_keys();
+            all_keys.insert(keys.begin(), keys.end());
+          }
+
+          // create a dict for the features
+          py::dict feature_dict;
+          py::list keys_list;
+          for (auto & key : all_keys) {
+            py::list l;
+            for (auto & ii : key) {
+              l.append(ii);
+            }
+            // convert it to a tuple
+            py::tuple t_key(l);
+            keys_list.append(t_key);
+            // feature_dict[t_key] = features;
+          }
+
+          int i_key{0};
+          for (auto & key : all_keys) {
```

Comment:
is there difference ?

Reply:
Not in behavior, but I find it clearer when iterating over multiple objects at the same time (until we get zip in c++!)

Reply:
I which we could but for it to be nice we should be using c++17.
Note that is this case I am iterating over one conceptual object that had to be split into 2 because of pybind requirements

Reply:
> I which we could but for it to be nice we should be using c++17.

Yep.

Do you think this is feasible? Looks like this means loosing support for gcc 5 & gcc 6 (https://en.cppreference.com/w/cpp/compiler_support), but gcc 7 should be mostly fine. Clang 4 already has almost full support.

Reply:
I am not sure it out weight the cost for this change. There are a few other nice things with variadic templates that have been added in cpp17 but again it does not seem to hold the cost. Do you know of other interesting features ?
We can ask next meeting anyway.

Reply:
I like the binding of multiple values `[auto a, auto b] = foo()`, `std::optional` is very nice, as well as `std::string_view`

Reply:
constexpr strings is nice indeed but not essential. 
Right optional and constexpr if would simplify most of our sfinae business.

Reply:
optional and string_view have C++98/C++11 compatible implementations: https://github.com/martinmoene/string-view-lite & https://github.com/martinmoene/optional-lite. But the rest of the standard library do not play well with theses.